mitodl/unified-ecommerce
# Add events system

### What are the relevant tickets?

https://github.com/mitodl/hq/issues/3158
https://github.com/mitodl/hq/issues/3427
(to a lesser extent) https://github.com/mitodl/unified-ecommerce/pull/93

### Description (What does it do?)

Prior PRs added the core system functionality to an extent that you can add items to the cart and pay for them. The next logical step is being able to trigger an action in the systems that own the products that were purchased so they can take the necessary action - e.g., enrolling the users that have paid for their courses. This adds a system to kick those events off.

This uses Pluggy to provide some hooks for events to be fired during the purchasing process. For now, this is pretty minimal and we only fire an event after the sale has been completed and after a refund has been processed. We will ideally need to fire more than this - we should also have a pre-sale hook, maybe some hooks for discounts, and some for financial assistance - but this gets the base system in to define those hooks.

### How can this be tested?

The application should not _externally_ work any differently. 

You can ensure that webhooks are being fired:
1. Set up some sort of bit bucket to provide a target for it to hit. I used https://github.com/rangle/catch-webhooks for this.
2. Update one of your integrated systems to hit the webhook. For local testing, I used `http://host.docker.internal:3003/webhook/sale-succeeded` and `http://host.docker.internal:3003/webhook/sale-refunded` - using the `host.docker.internal` hostname ensures that the app will try to connect to the Catch Webhooks copy running directly on my system.
3. Attempt to check out using a product that belongs to the integrated system. You should see the webhook being activated in the catch app.

You can also test multiple webhooks in a single cart. Repeat step 2 for another integrated system, then add products from both to your cart - you should see multiple webhooks be fired, and they should only contain line item data that pertains to their system. (For my testing, I added a `-2` to the webhook so I could differentiate between them easily.) Note that the total is the overall order total, not the per-system total.

### Additional Context

The data that is sent to the webhook is documented in the docs folder.

We don't have a way to refund currently except through the CyberSource EBC. (A future task will be to add some functionality to kick off refunds within UE, but we don't have that yet.)
